server/routes/auth.ts

```diff
@@ -641,4 +653,217 @@ authRoutes.post('/reset-password/:guid', async (req, res, next) => {
+        body: body,
+      });
+      return next({
+        status: 500,
+        message: 'Invalid token response.',
+      });
+    }
+
+    // Extract the ID token and access token
+    const { id_token: idToken, access_token: accessToken } = body as Extract<
+      typeof body,
+      { id_token: string; access_token: string }
+    >;
+
+    // Attempt to decode ID token jwt, catch and return any errors
+    const tryDecodeJwt = (): [IdTokenClaims | null, unknown] => {
+      try {
+        const decoded: IdTokenClaims = decodeJwt(idToken);
+        return [decoded, null];
+      } catch (error) {
+        return [null, error];
+      }
+    };
+    const [decoded, err] = tryDecodeJwt();
+
+    if (err != null) {
+      logger.info('Failed OIDC login attempt', {
+        cause: 'Invalid jwt',
+        ip: req.ip,
+        idToken: idToken,
+      });
+      return next({
+        status: 500,
+        message: 'Invalid jwt.',
+      });
+    }
+
+    // Merge claims from JWT with data from userinfo endpoint
+    const userInfo = await getOIDCUserInfo(wellKnownInfo, accessToken);
+    const fullUserInfo: FullUserInfo = { ...decoded, ...userInfo };
+
+    // Validate ID token jwt and user info
+    try {
+      const idTokenSchema = createIdTokenSchema({
+        oidcClientId: oidcClientId,
+        oidcDomain: oidcDomain,
+      });
+      await idTokenSchema.validate(fullUserInfo);
+    } catch (err) {
+      logger.info('Failed OIDC login attempt', {
+        cause: 'Invalid jwt or missing claims',
+        ip: req.ip,
+        idToken: idToken,
+        errorMessage: err.message,
+      });
+      return next({
+        status: 500,
+        message: `Validation failed: ${err.message}.`,
+      });
+    }
+
+    // Check that email is verified
+    if (!fullUserInfo.email_verified) {
+      logger.info('Failed OIDC login attempt', {
+        cause: 'Email not verified',
+        ip: req.ip,
+        email: fullUserInfo.email,
+      });
+      return next({
+        status: 500,
+        message: 'Email not verified.',
+      });
+    }
+
+    // Map email to user
+    const userRepository = getRepository(User);
+    let user = await userRepository.findOne({
+      where: { email: fullUserInfo.email },
+    });
+
+    // Map username to media server username
+    if (settings.main.oidcMatchUsername && !user) {
+      user = await userRepository.findOne({
+        where: [
+          { plexUsername: fullUserInfo.preferred_username },
+          { jellyfinUsername: fullUserInfo.preferred_username },
+        ],
+      });
+    }
+
+    // Create user if one doesn't already exist
+    if (!user) {
+      logger.info(`Creating user for ${fullUserInfo.email}`, {
+        ip: req.ip,
+        email: fullUserInfo.email,
+      });
+      const avatar =
+        fullUserInfo.picture ??
+        gravatarUrl(fullUserInfo.email, { default: 'mm', size: 200 });
+      user = new User({
+        avatar: avatar,
+        username: fullUserInfo.name,
```

Comment:
shouldn't you be using `fullUserInfo.preferred_username`? All systems involved in this process use the username field (similar to lign 823/824).